plugins/BEdita/Core/src/Model/Behavior/HistoryBehavior.php

```diff
@@ -0,0 +1,162 @@
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * See LICENSE.LGPL or <http://gnu.org/licenses/lgpl-3.0.html> for more details.
+ */
+
+namespace BEdita\Core\Model\Behavior;
+
+use BEdita\Core\State\CurrentApplication;
+use BEdita\Core\Utility\LoggedUser;
+use Cake\Datasource\EntityInterface;
+use Cake\Event\Event;
+use Cake\I18n\Time;
+use Cake\ORM\Behavior;
+use Cake\ORM\TableRegistry;
+
+/**
+ * History behavior
+ * Add history events on object modifications
+ *
+ * @since 4.1.0
+ */
+class HistoryBehavior extends Behavior
+{
+    /**
+     * {@inheritDoc}
+     */
+    protected $_defaultConfig = [
+        'table' => 'History',
+        'exclude' => ['type', 'id'],
+        'resource_type' => 'objects',
+    ];
+
+    /**
+     *  History table
+     *
+     * @var \Cake\ORM\Table
+     */
+    public $Table = null;
+
+    /**
+     * The changed properties.
+     * Array of visible properties changed by user
+     *
+     * @var array
+     */
+    public $changed = [];
+
+    /**
+     * {@inheritDoc}
+     */
+    public function initialize(array $config)
+    {
+        parent::initialize($config);
+        $table = $this->getConfig('table');
+        if (!empty($table)) {
+            $this->Table = TableRegistry::getTableLocator()->get($table);
+        }
+    }
+
+    /**
+     * Collect user changed properties.
+     *
+     * @param \Cake\Event\Event $event The event dispatched
+     * @param \ArrayObject $data The input data being saved
+     * @return void
+     */
+    public function beforeMarshal(Event $event, \ArrayObject $data)
+    {
+        $this->changed = $data->getArrayCopy();
+        $exclude = (array)$this->getConfig('exclude');
+        $this->changed = array_diff_key($this->changed, array_flip($exclude));
+    }
+
+    /**
+     * Save user changed properties in history table or other datasource.
+     *
+     * @param \Cake\Event\Event $event Fired event.
+     * @param \Cake\Datasource\EntityInterface $entity Entity data.
+     * @return void
+     */
+    public function afterSave(Event $event, EntityInterface $entity): void
+    {
+        if (empty($this->Table) || (empty($this->changed) && !$entity->isDirty('deleted'))) {
+            return;
+        }
+        $history = $this->historyEntity($entity);
+        $this->Table->saveOrFail($history);
+    }
+
+    /**
+     * Retrieve history event data.
+     *
+     * @param EntityInterface $entity Object entity.
+     * @return EntityInterface
+     */
+    protected function historyEntity(EntityInterface $entity): EntityInterface
+    {
+        /** @var \BEdita\Core\Model\Entity\History $history */
+        $history = $this->Table->newEntity();
+        $history->resource_id = $entity->get('id');
+        $history->resource_type = $this->getConfig('resource_type');
+        $history->application_id = CurrentApplication::getApplicationId();
```

Comment:
scrutinizer suggests a type check. maybe a cast is enough

```
$history->application_id = (int)CurrentApplication::getApplicationId();
```

Reply:
Actually the annotation `CurrentApplication::getApplicationId()` is wrong, should be `int|null` instead it's `Application|null` that's why Scrutinizer complains

Fixed annotation and pushed in this branch